### PR TITLE
Add nD support to image moments computation

### PR DESCRIPTION
**Pull request by jni**
## Description
Image moments can now be computed for images of any dimension. Many thanks to @jaimefrio for his optimization help on the mailing list for these functions. Reference: https://mail.python.org/pipermail/scikit-image/2017-April/005169.html

### CODE REVIEW

**Comment by jni** [comment]
All right @scikit-image/core! This one needs some decidin'. It turns out that our existing `moments` code uses xy coordinates, where for nD we would really want to use rc coordinates. This would normally be a straightforward deprecation (see code), but it turns out that image moments are used in a *lot* of regionprops. It's a deprecation *nightmare* to keep them all using xy. So my vote is to break the API and use rc throughout — at which point the deprecation might itself be worth skipping.

One big, massive hair in the ointment is orientation. Using the straightforward definition of the moments and of the orientation, it is now tilted by 90º relative to the usual xy orientation. This is the cause of the last remaining failing test. A horizontal line, which normally would have had orientation of 0, now has orientation of 90º. This makes sense when you think that the first axis (r) should be considered to have an orientation of 0.

Now, I'm among the most gung-ho supporters of rc coordinates (see above), but setting a vertical orientation as 0 is a bit much even for me. Should we maintain horizontal (right-facing) as the 0º orientation?

**Comment by soupault** [comment]
I think that `regionprops` is one of the most used features of `skimage`. That said, I think some kind of deprecation is still neccessary. How hard would it be to have 2 versions of `regionprops`: first - nD-rc, second - 2d-xy legacy? As we are following our internal agreement here, maybe we could be more aggresive and put a whole lot of deprecation warnings for 2d-xy one (forcing the users to switch to the newer version), and schedule the removal of it for 0.15?

:+1: to have orientation at 0º (to the right). Looks to me to be the most popular and intuitive approach.

**Comment by soupault** [comment]
Apparently, this PR fixes https://github.com/scikit-image/scikit-image/issues/2625 as a side-effect :).

**Comment by stefanv** [comment]
I agree on keeping the positive x-axis as 0 degrees.

As for deprecation, hrm.  Would it be super annoying to rename the current moments to moments_xy, and raise a warning each time moments are calculated that says "Please note the coordinate convention for this changes recently"---and drop that after 2 releases?

**Comment by stefanv** [comment]
@jni I'd like to round up this PR; do you have any thoughts on the xy issue mentioned above?

**Comment by jni** [comment]
@stefanv I'm just dissatisfied with the possible solutions for this... I find the inconsistency of 2D with 3D jarring. And, if we keep orientation 0 at horizontal, then no deprecation is needed — just a perpetual check that the input image is 2D, in which case we transpose the axes. Yuck. At any rate, I *don't* think we should transpose the axes for 3D.

There is also the question of whether we transpose the axes for the entire moments computation or just for the rotation. In the former case, this means that things like the centroid coordinates will need to be reversed when computing from the moments. In the latter, the moments and the orientation won't be consistent with each other.

In short, it's not a major puzzle why I have continued to procrastinate over this PR. =P

**Comment by kne42** [comment]
@jni I have adding nD contour handling to your function in jni/scikit-image#2 in response to your comment on #2703.

**Comment by stefanv** [comment]
@jni Yes, I agree that we should have one convention that scales for all dimensionalities

**Comment by jni** [comment]
@kne42 thanks for your PR! I haven't had a chance to review it yet because I'm travelling, but I will soon!

**Comment by soupault** [comment]
I think we should deliver this with 0.14. Let me know if you need any help.

**Comment by jni** [comment]
@scikit-image/core this should be ready for another round of review. I've taken the approach of deprecating the xy coordinate system and going all-in on the rc coordinate system (with a standard deprecation cycle). This means that orientation becomes 0 at vertical, ie vector (1, 0) has orientation 0, as in xy coordinates.

**Comment by jni** [comment]
I'm not sure what the story is with the Travis failures... There appear to be several problems! In the Python 3.4 build, there's a strange [Sphinx error](https://travis-ci.org/scikit-image/scikit-image/jobs/286464353#L5032):

```
embedding documentation hyperlinks for auto_examples... [  3%] plot_equalize.html
Exception occurred:
  File "/home/travis/venv/lib/python3.4/site-packages/sphinx_gallery/docs_resolv.py", line 185, in _get_link
    anchor = match[3]
IndexError: list index out of range
```

The Python 2.7 builds have actual test failures in the Normalized and Hu moments:

https://travis-ci.org/scikit-image/scikit-image/jobs/286464352#L4207

I'll log this as another instance of needing to drop 2.7 support 😛, but in the meantime we should probably figure out what's going on. Does anyone have any suggestions/ideas? Note that the very same tests pass in 3.x! Numerics are not supposed to be a source of differences between 2.x and 3.x! Grrrr.

**Comment by jni** [comment]
@scikit-image/core this is ready for review!

**Comment by jni** [comment]
I'll point out that the Travis failure is unrelated to the PR...

File: skimage/measure/_moments.py [hunk 1]
```diff
@@ -186,3 +219,78 @@ def moments_hu(nu):
...
+def inertia_tensor(image, mu=None):
+    """Compute the inertia tensor of the input image.
+
+    Parameters
+    ----------
+    image : array
+        The input image.
+    mu : array, optional
+        The pre-computed central moments of ``image``.
+    Returns
```
Comment (blink1073) [inline skimage/measure/_moments.py]: Need a blank line here?

Comment (jni) [reply skimage/measure/_moments.py]: @blink1073 fixed!

👋

File: skimage/measure/_moments.py [hunk 2]
```diff
@@ -186,3 +219,79 @@ def moments_hu(nu):
...
+    -------
+    T : array, shape ``(image.ndim, image.ndim)``
+        The inertia tensor of the input image.
+    """
+    if mu is None:
+        mu = moments_central(image)
+    mu0 = mu[(0,) * image.ndim]
+    result = np.zeros((image.ndim, image.ndim))
+
+    corners2 = tuple(2 * np.eye(image.ndim, dtype=int))
```
Comment (stefanv) [inline skimage/measure/_moments.py]: What's happening here?

File: skimage/measure/_moments.py [hunk 3]
```diff
@@ -186,3 +219,79 @@ def moments_hu(nu):
...
+        The centroid of the (nonzero) pixels in ``image``.
+    """
+    M = moments_central(image, center=(0,) * image.ndim, order=1)
+    center = M[tuple(np.eye(image.ndim, dtype=int))]
+    return center
+
+
+def inertia_tensor(image, mu=None):
+    """Compute the inertia tensor of the input image.
+
```
Comment (stefanv) [inline skimage/measure/_moments.py]: Please add an extended description around what is being computed here.

File: skimage/measure/_moments.py [hunk 4]
```diff
@@ -186,3 +219,79 @@ def moments_hu(nu):
...
+    Parameters
+    ----------
+    image : array
+        The input image.
+    mu : array, optional
+        The pre-computed central moments of ``image``.
```
Comment (stefanv) [inline skimage/measure/_moments.py]: Add some rationale for having this parameter.

Comment (stefanv) [reply skimage/measure/_moments.py]: And some examples on how to use mu.

File: skimage/measure/_moments.py [hunk 5]
```diff
@@ -186,3 +219,79 @@ def moments_hu(nu):
...
+        mu_index = np.zeros(image.ndim, dtype=int)
+        mu_index[list(dims)] = 1
+        result[dims] = -mu[tuple(mu_index)] / mu0
+        result.T[dims] = -mu[tuple(mu_index)] / mu0
+    return result
+
+
+def inertia_tensor_eigvals(image, mu=None, T=None):
+    """Compute the eigenvalues of the inertia tensor of the image.
+
```
Comment (stefanv) [inline skimage/measure/_moments.py]: Some more description / references?

File: skimage/measure/_regionprops.py [hunk 6]
```diff
@@ -13,6 +14,13 @@
 __all__ = ['regionprops', 'perimeter']
 
 
+XY_TO_RC_DEPRECATION_MESSAGE = (
```
Comment (stefanv) [inline skimage/measure/_regionprops.py]: Great! I like the pointer to a website where we have space to write.

File: skimage/measure/_regionprops.py [hunk 7]
```diff
@@ -107,6 +115,21 @@ def __init__(self, slice, label, label_image, intensity_image,
         self._cache_active = cache_active
         self._cache = {}
         self._ndim = label_image.ndim
+        # Note: in PR 2603, we added support for nD moments in regionprops.
```
Comment (stefanv) [inline skimage/measure/_regionprops.py]: Impressive deprecationeering!

File: skimage/measure/_regionprops.py [hunk 8]
```diff
@@ -107,6 +115,21 @@ def __init__(self, slice, label, label_image, intensity_image,
...
+        # Many properties used xy coordinates, instead of rc. This attribute
+        # helps with the deprecation process and should be removed in 0.16.
```
Comment (stefanv) [inline skimage/measure/_regionprops.py]: And should be added to the TODO list right now :)

File: skimage/measure/_regionprops.py [hunk 9]
```diff
@@ -221,88 +235,89 @@ def mean_intensity(self):
...
 
     @_cached
-    @only2d
     def moments(self):
-        return _moments.moments(self.image.astype(np.uint8), 3)
+        M = _moments.moments(self.image.astype(np.uint8), 3)
+        if not self._warned:
+            from warnings import warn
+            warn(XY_TO_RC_DEPRECATION_MESSAGE)
+            self.warned = True
```
Comment (stefanv) [inline skimage/measure/_regionprops.py]: This is how warnings work by default.  Anyway, not up to us to set this up; warn and let the user worry about filter settings.

**Comment by jni** [comment]
@stefanv I believe I've addressed your comments.

I think the docs build was fixed in #2851. I've rebased to test that theory. ;)

**Comment by stefanv** [comment]
A moment to celebrate!

**Comment by ahojnnes** [comment]
Infinitely many moments... great work.

**Comment by jni** [comment]
🎉🎉🎉

**Comment by soupault** [comment]
That is a huge step towards rc-regionprops! Thanks, @jni !
